### DISCUSSION TITLE
Race condition when animating Ajax Cart after item added

### DISCUSSION THREAD

**Discussion by james0r** [EvgeniyMukhamedjanov/liquid-ajax-cart#2] [Q&A] [answered]
I'm trying to animate my ajax cart into view using Alpine.js after item has been added via the product add form.

I'm suspecting that the ajax cart is being updated by LAC after the response is received and that's what is interfering with the animation and instead just showing the ajax cart without animating in.

here's my liquid-ajax-cart logic:

```
<script type="application/json" data-ajax-cart-initial-state>
  {{ cart | json }}
</script>

<script type="module">
  import { subscribeToCartAjaxRequests } from '{{ 'liquid-ajax-cart-v1.1.0.static.js' | asset_url }}';

  subscribeToCartAjaxRequests((requestState, subscribeToResult) => {

    if (requestState.requestType === 'add') {
      console.log('Item add request sent')

      subscribeToResult(requestState => {
        if (requestState.responseData?.ok) {

           window.Alpine.store('globals').minicartIsShowing = true
        }
      })
    }
    }
  });
</script>
```

If I wrap my ajax cart visibility state change in a setTimeout I get the desired animation. Is there a way I can listen for when the cart has been updated so i know when to begin animation? 

My minicart section --> [https://gist.github.com/james0r/cf256bf980b121466e7ce5d9b9e4905d](https://gist.github.com/james0r/cf256bf980b121466e7ce5d9b9e4905d
)

**Comment by james0r** [comment, marked as answer]
**Edit: Solved**
There were a couple issues at play here, one being that i was losing applied classes to the cart that was being swapped out by liquid ajax cart. The solution to this was to move my root level Alpine.js revealed ajax cart to a wrapper wrapping the root of LAC.

Then I faced the issue with LAC wanting to be the root element of the section so I ended up moving my Alpine.js ajax cart root into theme.liquid and wrapping my `{% section 'minicart' %}` in the Alpine.js ajax cart root. 

Looks like this: 

```
  <div 
      class="minicart-wrapper"
      x-data 
      x-show="$store.globals.minicartIsShowing" 
      x-on:click.outside="$store.globals.minicartIsShowing = false"
      x-transition:enter="transition ease-out duration-300"
      x-transition:enter-start="off-screen-right"
      x-transition:enter-end="right-50"
      x-transition:leave="transition ease-in duration-300"
      x-transition:leave-start="right-50"
      x-transition:leave-end="off-screen-right"
      x-cloak 
      :aria-hidden="$store.globals.minicartIsShowing ? 'false' : 'true'"
      aria-modal="true"
      role="dialog" 
      >
        {% section 'minicart' %}
    </div>
```

Now when I set my Alpine.js minicart visibility flag to true from subscribeToResult() callback, I keep my transitions and the cart updates properly.

**Reply by EvgeniyMukhamedjanov** [reply]
Yeah, the body of the section gets replaced with new HTML all the time maybe that's why Alpine couldn't understand what is going on.
Btw, why do you use Alpine instead of plain CSS classes ?

**Reply by james0r** [reply]
I am actually using plain CSS classes, they're applied in individual animation states similar to how the bigger framework animation libraries do it.

https://alpinejs.dev/directives/transition

In their example, they use Tailwind, I just decided to write my own classes instead to keep things light.

I'm really moving towards Alpine for toggling state and simple 2-way binding. Just feels much cleaner and as I reference back to a previous site it's obvious where things are handled.

Here's my playground store where I'm demo'ing this by the way. https://store2-0-playground.myshopify.com/ password playground

**Reply by EvgeniyMukhamedjanov** [reply]
Looks good! I saw you used item keys to show loading indicators for a specific line item.
But keep in mind that all the other line item's controls don't work when there is an Ajax request. I did it like this because it is much easier to handle.

**Reply by james0r** [reply]
> Looks good! I saw you used item keys to show loading indicators for a specific line item. But keep in mind that all the other line item's controls don't work when there is an Ajax request. I did it like this because it is much easier to handle.

What do you mean by other line item controls don't work? the remove button appears to be working and updating cart. I did have a problem because i was watching for change request to activate my loading indicator, but remove is also a change request. I just made a condition in my callbacks to look for a change with quantity > 0. More complete micro-interactions would call for me to have a separate loading indicator for the remove button. Was planning on getting to that.

**Reply by EvgeniyMukhamedjanov** [reply]
I meant when a user clicks for example the "plus" button and if they try to click any other button before the "Plus" button finished its work -- the other buttons will not react on user's clicks. One action per time.

**Reply by james0r** [reply]
oh i see what you mean, that's desirable isn't it?

**Reply by EvgeniyMukhamedjanov** [reply]
I did it intentionally because I didn't want to spend time to some edge cases that might happen. I don't think it brings any inconveniences since Shopify responds fast enough. Maybe in the future will think about it, but it is not critical for now.

**Reply by EvgeniyMukhamedjanov** [reply]
That is why in the docs I recommended to use the `.js-ajax-cart-request-in-progress` body class to make the ajax sections looks "disabled".

**Reply by EvgeniyMukhamedjanov** [reply]
Just for UX

**Reply by EvgeniyMukhamedjanov** [reply]
@james0r 
Just a small update. 
Now it is possible to apply the `data-ajax-cart-section` attribute not only to a root element.

```liquid
{% comment %} sections/my-cart.liquid {% endcomment %}

<div class="my-cart">
  <div class="my-cart__wrapper" data-ajax-cart-section >
    <!-- cart content -->
  </div>
</div> 

{% schema %} { "name": "My Cart" } {% endschema %}
```
In the above example only the `.my-cart__wrapper` will be updated, so you can safely apply any animation to the `.my-cart` element.

It is available since the v1.5.0.
